src/polyswarm_api/endpoint.py

```diff
@@ -0,0 +1,254 @@
+        self.timeout = timeout
+
+    def execute(self, request):
+        if 'timeout' not in request:
+            request['timeout'] = self.timeout
+        return self.session.request(**request)
+
+    def unauth_execute(self, request):
+        return self.unauth_session.request(**request)
+
+    def _get_engine_names(self, request):
+        return self.unauth_session.request(**request)
+
+    def __getattr__(self, name):
+        if name in self.__dict__:
+            return self.__dict__[name]
+        return self.execute
+
+    def _download_to_fh(self, req, fh):
+        raise NotImplementedError
+
+    def download(self, request):
+        request, fh = request
+        if isinstance(fh, str):
+            fh = open(fh, 'wb')
+
+        req = self.execute(request)
+        return self._download_to_fh(req, fh)
+
+    def download_archive(self, request):
+        request, fh = request
+        if isinstance(fh, str):
+            fh = open(fh, 'wb')
+
+        req = self.unauth_execute(request)
+        return self._download_to_fh(req, fh)
+
+
+class PolyswarmFuturesExecutor(PolyswarmRequestExecutor):
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES):
+        super(PolyswarmFuturesExecutor, self).__init__(key, timeout, retries, PolyswarmHTTPFutures)
+
+    def _download_to_fh(self, req, fh):
+        # this is unfortunately the cleanest way I think I can do this with requests-futures
+        # derived partially from https://github.com/ross/requests-futures/issues/54
+        def do_download(r, f):
+            for chunk in r.iter_content(chunk_size=const.DOWNLOAD_CHUNK_SIZE):
+                f.write(chunk)
+            return r
+        resp = req.result()
+        resp.raise_for_status()
+        return self.session.executor.submit(do_download, resp, fh)
+
+
+class PolyswarmSynchronousExecutor(PolyswarmRequestExecutor):
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES):
+        super(PolyswarmSynchronousExecutor, self).__init__(key, timeout, retries, PolyswarmHTTP)
+
+    def _download_to_fh(self, req, fh):
+        for chunk in req.iter_content(chunk_size=const.DOWNLOAD_CHUNK_SIZE):
+            fh.write(chunk)
+        return req
+
+
+class PolyswarmEndpoint(object):
+    """ This is the base class for PolyswarmEndpoint classes. Do not use directly. """
+    def __init__(self,  key, uri=const.DEFAULT_GLOBAL_API, community=const.DEFAULT_COMMUNITY,
+                 timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES,
+                 request_gen_cls=PolyswarmRequestGenerator, request_exec_cls=PolyswarmFuturesExecutor):
+        self.executor = request_exec_cls(key, timeout, retries)
+        self.generator = request_gen_cls(uri, community)
+
+    def __getattr__(self, name):
+        def endpoint_wrapper(*args, **kwargs):
+            return getattr(self.executor, name)(getattr(self.generator, name)(*args, **kwargs))
```

Comment:
Ok, this is the important magical method you are creating that makes using this easy. However, following what is happening here is not easy. For instance, when I look at this I think that the executor should have an attribute with the same name as the generator, but then `__getattr__()` is overwritten in the executor as well, which makes it really hard to understand the code.

Also, the executor don't really need to have their magical methods overwritten. Just make them regular methods. They will rarely be used outside of this class, and it make it much easier to understand what is happening. Readability counts.

This `getattr(self.executor, name)` will likely return `PolyswarmFuturesExecutor().execute` most of the time (if not all, I don't really see which other case it would not). You created a really nice interface here, use it directly as

```python
self.executor.execute(getattr(self.generator, name)(*args, **kwargs))
``` 

This is much more legible, you make the code more readable, you can remove `PolyswarmFuturesExecutor.__getattr__()` with no consequence (making it simpler) and it will work like a charm. It will also be easier for people to implement their own executors if the logic is more predictable. Overwriting magic methods is a tricky business and can lead to unexpected side-effects. I would only use it if it increases the usability significantly over being a bit more verbose.

There is one more point here. You are returning a closure at this point. I would let this be decided by the executor, meaning that I would execute the `execute()` right away and, if the executor is a `async` one, it will return a `future`, if not, it will return the response right away.

```python
    def __getattr__(self, name):
        return self.executor.execute(getattr(self.generator, name)(*args, **kwargs))
```

This is significantly simpler and you delegate the actual execution (or delayed execution, if the executor is a delayed one) to a decoupled part of the code. If I write an executor that puts things in a queue and I trigger it all afterwards I would expect to write some code like this:

```python
executor = QueuedExecutor()
endpoint = PolyswarmEndpoint(executor=executor)
futures = []
futures.append(endpoint.download(arg1))
futures.append(endpoint.submit(arg2))
futures.append(endpoint.rescan(arg3))
executor.run()
for future in futures:
    print(future.get_result())
```

While as it is, the usage would be

```python
executor = QueuedExecutor()
endpoint = PolyswarmEndpoint(executor=executor)
futures = []
futures.append(endpoint.download()(arg1))
futures.append(endpoint.submit()(arg2))
futures.append(endpoint.rescan()(arg3))
executor.run()
for future in futures:
    print(future.get_result())
```

Which makes for a strange empty function call there and more convoluted pattern. The user should expect that each call will return a result from the `execute()`, which can be a future or the result itself when using a immediate executor. It makes sense to build a pipeline of actions before executing them, so a user can choose another executor and then call `endpoint.executor.run()` or just `endpoint.run()` if you create an alias. Giving user control is a better pattern in my opinion here, and it makes things easier to understand and debug.

Reply:
> This getattr(self.executor, name) will likely return PolyswarmFuturesExecutor().execute most of the time (if not all, I don't really see which other case it would not). You created a really nice interface here, use it directly as

It does not in methods that it overrides; in this case, `download` and `download_archive` both wrap execute and are called via this method. This wasn't something that was done simply to further abuse `__getattr__`, though I agree it might be somewhat non-intuitive. Your example of delegating directly to execute does not work in all cases here. I could pass name in as well to the execute function, but is that the best way to move forward here?

Reply:
```
  def __getattr__(self, name):
        return self.executor.execute(getattr(self.generator, name)(*args, **kwargs))
```

This does not make sense to me. Where are `args` and `kwargs` coming from? Given we are inside `__getattr__`, we don't know yet how the user is going to call this (if at all). In the current code, this is the reason for creating the closure.

Reply:
```
executor = QueuedExecutor()
endpoint = PolyswarmEndpoint(executor=executor)
futures = []
futures.append(endpoint.download()(arg1))
futures.append(endpoint.submit()(arg2))
futures.append(endpoint.rescan()(arg3))
executor.run()
for future in futures:
    print(future.get_result())
```

This is absolutely not the current calling convention for endpoint. I think there may be some misunderstanding about how this functions. Each call already functions as what you've described as the "correct" behavior.

Reply:
You are correct, I got the usage wrong there, sorry about that. :grimacing: